### ISSUE TITLE
CPAL not finding devices (ALSA, Linux, Raspberry Pi) / CPAL does not support devices that only does recording?

### ISSUE THREAD

**Issue by JonasRSV** [RustAudio/cpal#515]
CPAL on raspberry pi does not find or pick any usable devices.

Printing all input devices 

```rust
    for device in cpal::default_host().input_devices().unwrap() {
        println!("Found {}", device.name().unwrap());
    }
```
Gives

```bash
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dsnoop.c:575:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
```

The output of arecord -L gives

```bash
null
    Discard all samples (playback) or generate zero samples (capture)
default:CARD=Camera
    USB2.0 Camera, USB Audio
    Default Audio Device
sysdefault:CARD=Camera
    USB2.0 Camera, USB Audio
    Default Audio Device
front:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    Front speakers
surround21:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    Direct sample mixing device
dsnoop:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    Direct sample snooping device
hw:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    Direct hardware device without any conversions
plughw:CARD=Camera,DEV=0
    USB2.0 Camera, USB Audio
    Hardware device with all software conversions
```
I have used alsa directly on the same raspberry pi and it works, what could be the reason why cpal does not find devices using alsa? :/ 

E.g this code

https://gist.github.com/albanpeignier/104902 can record fine with for example

```bash
gcc -o alsa-record-example -lasound simple_alsa.c && ./alsa-record-example default:CARD=Camera

audio interface opened
hw_params allocated
hw_params initialized
hw_params access setted
hw_params format setted
hw_params rate setted
hw_params channels setted
hw_params setted
hw_params freed
audio interface prepared
buffer allocated
read 0 done
read 1 done
read 2 done
read 3 done
read 4 done
read 5 done
read 6 done
read 7 done
read 8 done
read 9 done
buffer freed
audio interface closed
```

If I try to record with CPAL I get

```bash
A backend-specific error has occurred: ALSA function 'snd_pcm_open' failed with error 'ENOENT: No such file or directory'
```

If i print what device CPAL picks it says

```bash
Using device default
```

But that is curious since it found no device? :)

What would be next steps to get this to work with CPAL?

I am using CPAL for https://github.com/JonasRSV/Friday here is the specific code if it is relevant: https://github.com/JonasRSV/Friday/blob/main/friday/audio-io/src/lib.rs#L55 :)

**Comment by JonasRSV** [comment]
Using the rust crate alsa directly I am able to see my cards on the raspberry pi

```rust
        for card in alsa::card::Iter::new() {
            println!("Card {}", card.unwrap().get_name().unwrap());
            let iface = CString::new("pcm");

            for device in alsa::device_name::HintIter::new(Some(&card.unwrap()), &iface.unwrap()).unwrap() {
                println!("Device {}", device.name.unwrap());

            }
        }

```

Gives

```bash
Card bcm2835 Headphones
Device default:CARD=Headphones
Device sysdefault:CARD=Headphones
Device dmix:CARD=Headphones,DEV=0
Device dsnoop:CARD=Headphones,DEV=0
Device hw:CARD=Headphones,DEV=0
Device plughw:CARD=Headphones,DEV=0
Card USB2.0 Camera
Device default:CARD=Camera
Device sysdefault:CARD=Camera
Device front:CARD=Camera,DEV=0
Device surround21:CARD=Camera,DEV=0
Device surround40:CARD=Camera,DEV=0
Device surround41:CARD=Camera,DEV=0
Device surround50:CARD=Camera,DEV=0
Device surround51:CARD=Camera,DEV=0
Device surround71:CARD=Camera,DEV=0
Device iec958:CARD=Camera,DEV=0
Device dmix:CARD=Camera,DEV=0
Device dsnoop:CARD=Camera,DEV=0
Device hw:CARD=Camera,DEV=0
Device plughw:CARD=Camera,DEV=0
```

**Comment by JonasRSV** [comment]
Trying to print all devices with cpal

```rust
    for device in cpal::default_host().devices().unwrap() {
        println!("Found device {}", device.name().unwrap());
    }
```

Gives 

```bash
Found Host ALSA
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dsnoop.c:575:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
Found device default:CARD=Headphones
Found device sysdefault:CARD=Headphones
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Found device dmix:CARD=Headphones,DEV=0
ALSA lib pcm_dsnoop.c:575:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
Found device hw:CARD=Headphones,DEV=0
Found device plughw:CARD=Headphones,DEV=0
Using device default
```

I think maybe something is going wrong in the iterator

here: https://github.com/RustAudio/cpal/blob/master/src/host/alsa/enumerate.rs#L21 

because this loc https://github.com/RustAudio/cpal/blob/master/src/host/alsa/enumerate.rs#L13 contains the correct answer.

**Comment by JonasRSV** [comment]
Yes I think the problem is this: https://github.com/RustAudio/cpal/blob/master/src/host/alsa/enumerate.rs#L34

Does CPAL only work with devices that have only playback and recording? 

my device only have recording..

**Comment by JonasRSV** [comment]
So this works fine

```rust
        for device in alsa::device_name::HintIter::new_str(None, "pcm").unwrap() {
            let name = device.name.unwrap();
            let has_available_input = {
                let capture_handle =
                    alsa::pcm::PCM::new(&name, alsa::Direction::Capture, true);
                capture_handle.is_ok()
            };

            if has_available_input {
                println!("Is recording device {}", name);
            }
        }
```

```bash
Is recording device null
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
Is recording device default:CARD=Camera
Is recording device sysdefault:CARD=Camera
Is recording device front:CARD=Camera,DEV=0
Is recording device surround21:CARD=Camera,DEV=0
Is recording device surround40:CARD=Camera,DEV=0
Is recording device surround41:CARD=Camera,DEV=0
Is recording device surround50:CARD=Camera,DEV=0
Is recording device surround51:CARD=Camera,DEV=0
Is recording device surround71:CARD=Camera,DEV=0
Is recording device iec958:CARD=Camera,DEV=0
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Is recording device dsnoop:CARD=Camera,DEV=0
Is recording device hw:CARD=Camera,DEV=0
Is recording device plughw:CARD=Camera,DEV=0
```

**Comment by JonasRSV** [comment]
But this does not

```rust
        for device in alsa::device_name::HintIter::new_str(None, "pcm").unwrap() {
            let name = device.name.unwrap();
            let io = device.direction;

            if let Some(io) = io {
                if io != alsa::Direction::Playback {
                    continue;
                }
            }
            let has_available_input = {
                let capture_handle =
                    alsa::pcm::PCM::new(&name, alsa::Direction::Capture, true);
                capture_handle.is_ok()
            };

            if has_available_input {
                println!("Is recording device {}", name);
            }
        }

```

```bash
Is recording device null
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
```

**Comment by JonasRSV** [comment]
Printing the io

```rust
        for device in alsa::device_name::HintIter::new_str(None, "pcm").unwrap() {
            let name = device.name.unwrap();
            let io = device.direction;

            if let Some(io) = io {
                println!("device {} io: {:?}", name, io);
                if io != alsa::Direction::Playback {
                    continue;
                }
            }
            let has_available_input = {
                let capture_handle =
                    alsa::pcm::PCM::new(&name, alsa::Direction::Capture, true);
                capture_handle.is_ok()
            };

            if has_available_input {
                println!("Is recording device {}", name);
            }
        }
```

```bash
Is recording device null
device default:CARD=Headphones io: Playback
device sysdefault:CARD=Headphones io: Playback
device dmix:CARD=Headphones,DEV=0 io: Playback
ALSA lib pcm_dmix.c:1043:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
device dsnoop:CARD=Headphones,DEV=0 io: Playback
ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
device hw:CARD=Headphones,DEV=0 io: Playback
device plughw:CARD=Headphones,DEV=0 io: Playback
device default:CARD=Camera io: Capture
device sysdefault:CARD=Camera io: Capture
device front:CARD=Camera,DEV=0 io: Capture
device surround21:CARD=Camera,DEV=0 io: Capture
device surround40:CARD=Camera,DEV=0 io: Capture
device surround41:CARD=Camera,DEV=0 io: Capture
device surround50:CARD=Camera,DEV=0 io: Capture
device surround51:CARD=Camera,DEV=0 io: Capture
device surround71:CARD=Camera,DEV=0 io: Capture
device iec958:CARD=Camera,DEV=0 io: Capture
device dmix:CARD=Camera,DEV=0 io: Capture
device dsnoop:CARD=Camera,DEV=0 io: Capture
device hw:CARD=Camera,DEV=0 io: Capture
device plughw:CARD=Camera,DEV=0 io: Capture
```

So If I understand it correctly, the iterator only lets through devices that have recording and playback capabilities, or only playback. But not only recording?

Can't that if-statement just be dropped or is there some problem with having only recording? :)

**Comment by mitchmindtree** [comment]
It sounds like this is a bug - it's very possible that most folks using ALSA have gotten lucky so far by using devices that happen to support both capture *and* playback. Thanks for the detailed issue!

A PR (with all examples tested) would certainly be very welcome.

**Comment by JonasRSV** [comment]
Yes i'd be happy to look into creating a PR, I currently have exams coming up though so it won't be until the end of December.

I think most people won't be affected by this, because CPAL automatically picks the 'default' device, and often the 'default' device is the one that the users wants used. But on raspberry PI the default device does not work since the PI has no sound card. 

I worked around this by changing the default device on the PI to be the device I wanted, so to clarify, CPAL works with 'record only' devices. But it does not detect them, using ALSA at least. So as long as the 'record only' device is the 'default' device it is no problem.

But yes not having to change the default device and have CPAL be able to see 'record  only' devices would be nice and i'll look into it later in December :)

**Comment by alexmoon** [comment]
I've included a fix for this in #506 since I was refactoring that method anyway.
